### PR TITLE
Document agent detailed upgrade states

### PR DESCRIPTION
**Pull request by kilfoyle**
**Wait on https://github.com/elastic/kibana/pull/168638 before merging.**

---

This documents the very nice changes via https://github.com/elastic/kibana/pull/167539 so that detailed Elastic Agent upgrade states are now visible in the Fleet UI. This applies to 8.12 and higher.

Closes: #330

Docs preview: [View upgrade status](https://ingest-docs_596.docs-preview.app.elstc.co/guide/en/fleet/master/upgrade-elastic-agent.html#view-upgrade-status)

### CODE REVIEW

**Comment by jillguyonnet** [comment]
@kilfoyle I noticed there was an upgrade state missing from the screenshot I posted in the PR description. Can you use this one instead:

Screenshot 2023-10-30 at 15 31 26

Let me know if you require an image with better quality.

File: docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc
Comment (jillguyonnet) [inline docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc:148]: How about listing the upgrade states in the order they happen? These can be found [here](https://github.com/elastic/elastic-agent/issues/3119).

File: docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc
Comment (jillguyonnet) [inline docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc:160]: The target version is currently 8.12, although that might change.

**Comment by kilfoyle** [comment]
Thanks @jillguyonnet. I've updated the graphic. It looks nice and clear in the output. I've also reordered the status descriptions to be in chronological order and noted that the upgrade state tooltips are available for 8.12 and up.

**Comment by jillguyonnet** [comment]
Thanks for the changes @kilfoyle!

I hadn't read the paragraph that mentions the version properly yesterday, sorry about that. After another read, I think there are a few things we can clarify:
- Agents on 8.12 (TBC) and higher will have upgrade status and the information icon, so we should mention the version at the beginning of the paragraph.
- Since the user might have agents on older versions with no upgrade details, it would be good to mention that as well.
- We should still mention that upgrading agents will have an `Updating` state, regardless of version (by the way, I'd like to understand why it's not called `Upgrading`, but that's a different question).

With that in mind, what do you think of something along the following lines:
- At the beginning of the `View upgrade status` section, leave the original sentence "On the **Agents** tab, when you trigger an upgrade, agents that are upgrading have the status `Updating` until the upgrade is complete, then the status
 changes back to `Healthy`." since that applies to all agents.
- Then, say something like "Agents on version 8.12 and higher that are currently  upgrading additionally show a detailed upgrade status indicator...".
- Remove the version mention in the paragraph that mentions the information icon, so something like "You can hover your cursor..."
- Add a paragraph before "When upgrading many agents..." that says something like "Upgrading agents on versions older than 8.12 do not provide upgrade details" and add a screenshot like [this one](https://user-images.githubusercontent.com/23701614/272950148-353f5942-e74f-468a-9430-895b7dbe19e5.png) (it's not the best because of that weird warning icon, so I can do a new one if you want).

**Comment by jillguyonnet** [comment]
One quick additional comment: since we're waiting on the Elastic Agent implementation to be complete to know the target version (aiming for 8.12, but it might change), this should not be merged until we are ready to release. One way to know is when [this PR](https://github.com/elastic/kibana/pull/168638) will be merged.

**Comment by jillguyonnet** [comment]
New screenshot for an upgrading agent on an older version:

Screenshot 2023-10-31 at 09 06 37

**Comment by kilfoyle** [comment]
Thanks @jillguyonnet! These are great ideas. I've made the changes just as you've suggested. For the new graphic, I zoomed in a little bit as otherwise the text doesn't display very clearly when rendered in the docs.

The [preview page](https://ingest-docs_596.docs-preview.app.elstc.co/guide/en/fleet/master/upgrade-elastic-agent.html#view-upgrade-status) is up to date.

I put a note for myself in the description not to merge this until the tooltips PR is merged.

**Review by jillguyonnet — APPROVED** [review]
Nice, thank you David!
Let's wait until the implementation is completed and confirm that the version is still 8.12, then we'll be good to go.

File: docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc
Comment (ycombinator) [inline docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc:149]: I'd omit the second clause in the description because it isn't accurate for scheduled upgrades.

```suggestion
| Upgrade requested | {agent} has received the upgrade notice from {fleet}.
```

File: docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc
Comment (ycombinator) [inline docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc:155]: Typo?

```suggestion
| Upgrade monitoring | The newly upgraded {agent} has started and is being monitored for errors.
```

File: docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc
Comment (ycombinator) [inline docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc:156]: Consistency with terms used in the "Upgrade replacing" description.

```suggestion
| Upgrade rolled back | The upgrade was unsuccessful. {agent} is being rolled back to the former, pre-upgrade version.
```

File: docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc
Comment (ycombinator) [inline docs/en/ingest-management/fleet/upgrade-elastic-agent.asciidoc:169]: Perhaps less ambiguous?

```suggestion
Note that when you upgrade agents from versions below 8.12, the upgrade details are not provided.
```

**Comment by ycombinator** [comment]
As it stands, upgrade details are conveyed by Elastic Agent to Fleet whenever Elastic Agent checks in with Fleet Server.  These check-ins happen on a fairly infrequent basis.  This is by design, as we don't want to overwhelm Fleet Server, particularly if several Agents are all being upgraded at once.  We do [plan](https://github.com/elastic/elastic-agent/issues/3119#issuecomment-1745128570) to do some performance testing to try and increase the frequency of check-ins during the upgrade process.

As such, as things stand, a user may not see an Agent in the midst of an upgrade transitioning through all the states.  In fact, if the upgrade goes smoothly without an issues and the Agent is able to download the new artifact fast enough, the user may simply see the "upgrade requested" or "upgrade scheduled" states, followed by the upgrade being completed, in which case they'll see the new Agent version in the UI and nothing where the upgrade state label is shown.

I'm not sure if this "best case" behavior is something we should mention in the documentation (without the implementation details from the first paragraph of this comment) or leave it out?  Or perhaps we could allude to it in a more subtle manner, by stating that one of the purposes of upgrade details is to understand the progress of an ongoing upgrade, particularly if it is stuck or appears to be stuck at any point?

**Comment by kilfoyle** [comment]
That's a great point Shaunak! Do you think this new paragraph would help?
    Screenshot 2023-11-20 at 3 17 53 PM

Also, I've added in your other suggestions via [1716dbe](https://github.com/elastic/ingest-docs/commit/1716dbe2e801060a5bde7842bcfd5502194adb91)

**Comment by kilfoyle** [comment]
By the way, the security on this repo has been updated so new commits automatically dismiss previous approvals.

**Comment by ycombinator** [comment]
> Do you think this new paragraph would help?

Yes, I think this new paragraph should help address any user concerns around not seeing all the upgrade states while an Agent is being upgraded.

Nitpick: the paragraph appears above the table where we explain the different upgrade states.  Also, before this paragraph was added, the first occurrence of "upgrade states" was in the table header.  Should we move the paragraph below the table, after the user has been introduced to concept of upgrade states?

**Review by ycombinator — APPROVED** [review]
Left a nitpick suggestion in https://github.com/elastic/ingest-docs/pull/596#issuecomment-1819789368.  Otherwise, LGTM.

**Comment by kilfoyle** [comment]
> Should we move the paragraph below the table, after the user has been introduced to concept of upgrade states?

Definitely. Done via [dc58b8d](https://github.com/elastic/ingest-docs/pull/596/commits/dc58b8d49cbd7ad9f43865d885ef41807833f948)

**Review by jillguyonnet — APPROVED** [review]
Nice changes, thank you 👍 
https://github.com/elastic/kibana/pull/168638 is now merged so this is OK for 8.12.

**Review by ycombinator — APPROVED** [review]
LGTM — nice to have such excellent documentation around this feature!

**Review by lcawl — APPROVED** [review]

**Pull request merged**
